README.md
# OSC controlled simple clock
* Clock binary builds: [![pipeline status](https://gitlab.com/Depili/clock-8001/badges/master/pipeline.svg)](https://gitlab.com/Depili/clock-8001/commits/master)
* Clock image builds: [![pipeline status](https://gitlab.com/Depili/buildroot-clock-8001/badges/master/pipeline.svg)](https://gitlab.com/Depili/buildroot-clock-8001/commits/master)

Support clock-8001 development by paypal: [![](https://www.paypalobjects.com/en_US/i/btn/btn_donate_LG.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=XUMXUL5RX5MWJ&currency_code=EUR)


This is a simplistic clock written in go that can be used either as a video out with SDL or as a dedicated clock buidt with a 32x32 pixel hub75 led matrix and a ring of 60 addressable leds.

The README has been written for the version 4.x.x clocks, ie. the gitlab.com/Depili/clock-8001/v4 go module.

The clock can be controlled with the depili-clock-8001 companion module: https://github.com/bitfocus/companion-module-depili-clock-8001 Module version 5.0.0 is the first to implement V4 clock API.

Features and configuration in greated detail can be found in the [getting started guide in wiki](https://gitlab.com/Depili/clock-8001/-/wikis/Getting-started-on-clock-8001-version-4).

Developed in co-operation with Daniel Richert and with a grant from FUUG - Finnish Unix User Group.

## Ready made raspberry pi images

SD-card images for raspberry pi can be found at https://kissa.depili.fi/clock-8001/images

Clock-8001 no longer has multiple different images, they have all been consolidated to one unified image which uses `enable_` files to activate various parts of the clock system as desired. The new images are named as `clock-800-unified-<version>.img`

The images support raspberry pi 2B / 3B / 3B+ boards. They need at least 64Mb SD-cards. Write them to the card like any other raspberry pi sd-card image.

The image tries to get a dhcp address on wired ethernet and also brings up a virtual interface eth0:1 with static ip (default 192.168.10.245 with 255.255.255.0 netmask).

### Customizing the images

You can place the following files on the sd-card FAT partition to customize the installation:
* `clock.ini` main clock configuration file that is used by the default `clock_cmd.sh`
* `hostname` to change the hostname used by the clock, it is available with "hostname.local" for bonjour / mDNS requests
* `interfaces` a replacement for /etc/network/interfaces for custom network configuration
* `ntp.conf` for custom ntp server configuration
* `config.sys` the normal raspberry pi boot configuration for changing video modes etc.
* `sdl-clock` to update the clock binary with this file
* `clock_cmd.sh` is the command line for the clock, it should start with `/root/sdl-clock ` and be followed by any command line parameters you wish to use for the clock.
* `clock_bridge` to update the clock bridge binary file
* `clock_bridge_cmd.sh` to update the clock bridge command line. It should start with `/root/clock-bridge` and be followed by any command line paramaters for the bridge.
* `enable_clock` delete this file and the main clock will not be active
* `enable_bridge` delete this file and the mitti / millumin osc bridge will not be active
* `enable_ssh` delete this file and remote ssh logins to the raspberry pi will not be allowed
* `enable_ltc` delete this file and the LTC audio -> OSC functionality will not be active

### Hyperpixel4 displays

The image supports both the retangular and square hyperpixel4 displays from Pimoroni. To use them you need to rename either `config.txt.hp4_square` or `config.txt.hp4_rect` to `config.txt`

For the rectangular display you need to also modify `clock_cmd.sh` to contain:
```
/hyperpixel4_rect/hyperpixel4-init
/root/sdl-clock -C /boot/clock.ini
```
So that the display is initialized on boot.

### Web configuration interface

The new unified images have a web configuration interface for the clock settings. You can access this interface by pointing your browser to the address of the clock. The default username is `admin` and the default password is `clockwork`. You should change them from the interface or the clock.ini file.

## sdl-clock - Output the clock to hdmi on the raspberry pi

You can build the clock binary with `go get gitlab.com/Depili/clock-8001/cmd/sdl_clock`. Compiling requires SDL 2 and SDL_GFX 2 libraries. On the raspberry pi the default libraries shipped with rasbian will only output data to X11 window, so for full screen dedicated clock you need to compile the SDL libraries from source. For compiling use `./configure --host=armv7l-raspberry-linux-gnueabihf --disable-pulseaudio --disable-esd --disable-video-mir --disable-video-wayland --disable-video-x11 --disable-video-opengl` for config flags.

### Precompiled binaries

* Latest from git master: [sdl-clock](https://gitlab.com/Depili/clock-8001/-/jobs/artifacts/master/raw/sdl-clock?job=build)
* Testing builds: https://kissa.depili.fi/clock-8001/testing/
* Tagged releases: https://kissa.depili.fi/clock-8001/releases/

### Command line parameters
```
Usage:
  sdl-clock [OPTIONS]

Application Options:
  -C, --config=                                                read config from a file
      --face=[round|dual-round|small|text|single|countdown]    Select the clock face to use (default: round)
      --debug                                                  Enable debug output
      --http-port=                                             Port to listen on for the http configuration interface (default: :8080)
      --disable-http                                           Disable the web configuration interface
      --http-user=                                             Username for web configuration (default: admin)
      --http-password=                                         Password for web configuration interface (default: clockwork)
      --dump-config                                            Write configuration to stdout and exit
      --defaults                                               load defaults
      --no-ar-correction                                       Do not try to detect official raspberry pi display and correct it's aspect
                                                               ratio
      --background=                                            Background image file location.
      --background-path=                                       path to load OSC backgrounds from (default: /boot)
      --background-color=                                      Background color, used if no background image is supplied (default: #000000)
      --flash=                                                 Flashing interval when countdown reached zero (ms), 0 disables (default: 500)
      --osc-listen=                                            Address to listen for incoming osc messages (default: 0.0.0.0:1245)
  -d, --timeout=                                               Timeout for OSC message updates in milliseconds (default: 1000)
  -o, --osc-dest=                                              Address to send OSC feedback to (default: 255.255.255.255:1245)
      --disable-osc                                            Disable OSC control and feedback
      --disable-feedback                                       Disable OSC feedback
      --disable-ltc                                            Disable LTC display mode
      --ltc-seconds                                            Show seconds on the ring in LTC mode
      --udp-time=[off|send|receive]                            Stagetimer2 UDP protocol support (default: receive)
      --udp-timer-1=                                           Timer to send as UDP timer 1 (port 36700) (default: 1)
      --udp-timer-2=                                           Timer to send as UDP timer 2 (port 36701) (default: 2)
      --ltc-follow                                             Continue on internal clock if LTC signal is lost. If unset display will blank
                                                               when signal is gone.
      --format-12h                                             Use 12 hour format for time-of-day display
      --mitti=                                                 Counter number for Mitti OSC feedback (default: 8)
      --millumin=                                              Counter number for Millumin OSC feedback (default: 9)
      --millumin-ignore-layer=REGEXP                           Ignore matching millumin layers (case-insensitive regexp) (default: ignore)
      --info-timer=                                            Show clock status for x seconds on startup (default: 30)
  -F, --font=                                                  Font for event name (default: fonts/7x13.bdf)
      --text-color=                                            Color for round clock text (default: #FF8000)
      --static-color=                                          Color for round clock static circles (default: #505000)
      --second-color=                                          Color for round clock second circles (default: #C80000)
      --countdown-color=                                       Color for round clock second circles (default: #FF0000)
      --number-font=                                           Font for text clock face numbers (default: Copse-Regular.ttf)
      --label-font=                                            Font for text clock face labels (default: RobotoMono-VariableFont_wght.ttf)
      --icon-font=                                             Font for text clock face icons (default: MaterialIcons-Regular.ttf)
      --row1-color=                                            Color for text clock row 1 (default: #FF8000)
      --row1-alpha=                                            Alpha channel for text clock row 1 (default: 255)
      --row2-color=                                            Color for text clock row 2 (default: #FF8000)
      --row2-alpha=                                            Alpha channel for text clock row 2 (default: 255)
      --row3-color=                                            Color for text clock row 3 (default: #FF8000)
      --row3-alpha=                                            Alpha channel for text clock row 3 (default: 255)
      --label-color=                                           Color for text clock labels (default: #FF8000)
      --label-alpha=                                           Alpha channel for label text color (default: 255)
      --timer-bg-color=                                        Color for optional timer background box (default: #202020)
      --timer-bg-alpha=                                        Alpha channel for timer background boxes (default: 255)
      --label-bg-color=                                        Color for optional label background box (default: #202020)
      --label-bg-alpha=                                        Alpha channel for label bacground boxes (default: 255)
      --draw-boxes                                             Draw the container boxes for timers
      --numbers-size=
      --font-path=                                             Path for loading font choices into web config (default: .)
      --countdown-target=

1st clock display source:
      --source1.text=                                          Title text for the time source
      --source1.counter=                                       Counter number to associate with this source, leave empty to disable it as a
                                                               suorce (default: 0)
      --source1.ltc                                            Enable LTC as a source
      --source1.timer                                          Enable timer counter as a source
      --source1.tod                                            Enable time-of-day as a source
      --source1.timezone=                                      Time zone to use for ToD display (default: Europe/Helsinki)
      --source1.hidden                                         Hide this time source

2nd clock display source:
      --source2.text=                                          Title text for the time source
      --source2.counter=                                       Counter number to associate with this source, leave empty to disable it as a
                                                               suorce (default: 0)
      --source2.ltc                                            Enable LTC as a source
      --source2.timer                                          Enable timer counter as a source
      --source2.tod                                            Enable time-of-day as a source
      --source2.timezone=                                      Time zone to use for ToD display (default: Europe/Helsinki)
      --source2.hidden                                         Hide this time source

3rd clock display source:
      --source3.text=                                          Title text for the time source

CHANGELOG.md
## Version 4.6.0
* Features:
  * 144x144px round clock face
  * 288x144px text clock face
  * Improved signal color handling, You can now override the automation set colors via osc commands
    * If enabled, automation does one color change when timer thresholds are reached, but doesn't prevent manually setting the signal colors.


## Version 4.5.2
* Bugfix: /clock/signal/* OSC command was expecting 4 parameters

## Version 4.5.1
* Bugfixes:
  * Config generation for overtime colors
  * Round clock and overtime visibility modes

## Version 4.5.0
* New: Customizable overtime behaviour
  * Handling of overtime countdowns can now be customized
  * Readout can be chosen to be either: show zeros, show blank, continue counting up
  * The extra visibility can be chosen to be: none, blink, change background color or change both background and blink

## Version 4.4.1
* Bugfix: Load the SPI kernel modules in the clock image for unicorn hats

## Version 4.4.0
* New: Initial hardware signal light support
  * Currently only pimoroni Unicorn HD and Ubercorn hats are supported
  * You can control the color via OSC by groups, or the light color can follow source 1

## Version 4.3.1
* Bugfix for the signal colors

## Version 4.3.0
* New: Color signals for timers
  * `/clock/timer/*/signal` OSC command
  * Optional automation for the signal colors based on timer thresholds

## Version 4.2.0
* New:
  * /clock/flash command, it flashes the screen white for 200ms

## Version 4.1.1
* Fix version information stamping in build automation

## Version 4.1.0
* New:
  * Support for hyperpixel4 square displays
  * Documentation for the displays in README.md

## Version 4.0.3
* Bugfixes:
  * Unhide sources when starting a targeted count

## Version 4.0.2
* Bugfixes:
  * Hide behaviour of the different osc commands to be uniform
  * Starting a timer to given target while clock is paused

## Version 4.0.1
* Fixed hyperpixel4-square configuration

## Version 4.0.0
### New feature highlights

* Support for different clock faces
  * Single round clock
  * Dual round clocks
  * Text clocks
    * 3 and 1 clocks per screen
* New concept of sources and timers
  * A clock face displays a clock source, which can be associated to a timer and different sources for the time, eg. LTC
  * Timers can be used on multiple clock setups
* HTTP configuration editor with input validation
  * import / export for the whole configuration file
* New OSC commands, see osc.md
  * New companion module version supporting the V4 commands and feedback
  * Background selection
  * Timers targeting a certain time-of-day
  * Sending of text messages with custom text and background colors
  * Setting of clock source labels
* Info overlay shown on startup containing version, ip and port information
* Support of LTC input via usb-soundcard / hifiberry dac+ adc pro hat
* Mitti and millumin message processing completely built in
  * You can choose which timers are the destinations of the playback state
* Support for sending and receiving timers as Interspace Industries Countdown2 UDP messages for display units
  * Also supported by StageTimer2 and Irisdown

### Breaking changes

* clock configuration file contents have been changed.
* OSC commands have been changed, compatibility with v3 commands has been kept as close as possible
  * V3 commands will be dropped at a later date
* OSC feedback has been changed
* Command line arguments have been changed
* New dependencies; sdl-ttf, sdl-image, ttf fonts for text rendering

## Version 3.16.3
* BUGFIX: Do not unpause counters when started

## Version 3.16.2
* BUGFIX: Timer pause, resume and turning clock off
* BUGFIX: /clock/countup/modify direction flip to be more natural

## Version 3.16.1
* Fix a linter error preventing automatic builds

## Version 3.16.0
* Start refactoring internal clock engine to implement new clock faces
* Add /clock/countup/modify OSC command
* Add /boot/config.txt editing to the web configuration

## Version 3.15.0
* Add experimental support for Interspace Industries USB AiO soundcard for LTC to the generated images

## Version 3.14.1
* BUGFIX: Clock scaling on the official 7" display

## Version 3.14.0
* Experimental background image support

## Version 3.13.0
* Add an option to disable detection and aspect ratio correction of official raspberry pi displays
* BUGFIX: setting 12h format from the http interface didn't work
* Internal refactoring and code cleanup

## Version 3.12.1
* BUGFIX: missing colon from default clock.ini for port 80

## Version 3.12.0
* Add support for 12 hour format on time-of-day display

## Version 3.11.1
* LTC Bugfix, we had a wrong audio device in the default configuration

## Version 3.11.0
* LTC added options:
  * Toggle between frames and seconds on the clock ring
  * Toggle for loss of signal handling; either blank the clock or continue on internal time
  * Toggle for disabling the LTC handling
* Images
  * Build raspberry pi 2 compatible images
  * Use config file generated by the build automation

## Version 3.10.2
* LTC reception & Hifiberry ADC+ DAC Pro
  * Bugfix on sdl-clock to display the hour part of LTC timestamp
  * Added configuration for the hifiberry to the sd-card image

## Version 3.10.1
* Little more black space between dual clock text display "pixels"


## Version 3.10.0
* Add /clock/dual/text OSC message for setting a text field on dual clock mode

## Version 3.9.0
* Build automation:
  * Build one unified image
* Clock:
  * Web configuration interface
  * Dual-clock mode

## Version 3.8.2
* Build automation improvements
  * Read version information from runtime/debug.ReadBuildInfo()
  * Include pimoroni hyperpixel4 square display support

## Version 3.8.0
* multi-clock: Implement new display of up to 40 clocks
* config file support: use -C to read config flags from a file
* osx support: Remove unused rpi gpio stuff from sdl-clock and handle sdl2 events

## Version 3.7.1
* clock-bridge:

clock/engine.go
package clock

import (
	"fmt"
	"github.com/desertbit/timer"
	"github.com/hypebeast/go-osc/osc"
	"gitlab.com/Depili/clock-8001/debug"
	"log"
	"math"
	"os"
	"os/exec"
	"regexp"
	"time"
)

// Version is the current clock engine version
const Version = "3.0.0"

// Will get overridden by ldflags in Makefile
var gitCommit = "Unknown"
var gitTag = "v0.0.0"

// EngineOptions contains all common options for clock.Engines
type EngineOptions struct {
	Flash           int    `long:"flash" description:"Flashing interval when countdown reached zero (ms), 0 disables" default:"500"`
	Timezone        string `short:"t" long:"local-time" description:"Local timezone" default:"Europe/Helsinki"`
	ListenAddr      string `long:"osc-listen" description:"Address to listen for incoming osc messages" default:"0.0.0.0:1245"`
	Timeout         int    `short:"d" long:"timeout" description:"Timeout for OSC message updates in milliseconds" default:"1000"`
	Connect         string `short:"o" long:"osc-dest" description:"Address to send OSC feedback to" default:"255.255.255.255:1245"`
	CountdownRed    uint8  `long:"cd-red" description:"Red component of secondary countdown color" default:"255"`
	CountdownGreen  uint8  `long:"cd-green" description:"Green component of secondary countdown color" default:"0"`
	CountdownBlue   uint8  `long:"cd-blue" description:"Blue component of secondary countdown color" default:"0"`
	DisableOSC      bool   `long:"disable-osc" description:"Disable OSC control and feedback"`
	DisableFeedback bool   `long:"disable-feedback" description:"Disable OSC feedback"`
}

// Clock engine state constants
const (
	Normal    = iota // Display current time
	Countdown = iota // Display countdown timer only
	Countup   = iota // Count time up
	Off       = iota // (Mostly) blank screen
	Paused    = iota // Paused countdown timer(s)
)

type countdownData struct {
	target   time.Time     // Target timestamp for main countdown
	duration time.Duration // Total duration of main countdown, used to scale the leds
	left     time.Duration // Duration left when paused
	active   bool
}

// Engine contains the state machine for clock-8001
type Engine struct {
	timeZone       *time.Location // Time zone, initialized from options
	mode           int            // Main display mode
	countdown      *countdownData
	countdown2     *countdownData
	countup        *countdownData
	paused         bool
	Hours          string
	Minutes        string
	Seconds        string
	displaySeconds bool
	Tally          string
	TallyRed       uint8
	TallyGreen     uint8
	TallyBlue      uint8
	cd2Red         uint8
	cd2Green       uint8
	cd2Blue        uint8
	Leds           int
	Dots           bool
	flashLeds      bool
	flasher        *time.Ticker
	clockServer    *Server
	oscServer      osc.Server
	timeout        time.Duration        // Timeout for osc tally events
	oscTally       bool                 // Tally text was from osc event
	oscDests       *feedbackDestination // udp connections to send osc feedback to
	initialized    bool                 // Show version on startup until ntp synced or receiving OSC control
}

// MakeEngine creates a clock engine
func MakeEngine(options *EngineOptions) (*Engine, error) {
	var engine = Engine{
		mode:           Normal,
		Hours:          "",
		Minutes:        "",
		Seconds:        "",
		displaySeconds: true,
		Leds:           0,
		flashLeds:      true,
		Dots:           true,
		oscTally:       false,
		paused:         false,
		timeout:        time.Duration(options.Timeout) * time.Millisecond,
		cd2Red:         options.CountdownRed,
		cd2Green:       options.CountdownGreen,
		cd2Blue:        options.CountdownBlue,
		initialized:    false,
		oscDests:       nil,
	}

	log.Printf("Clock-8001 engine version %s git: %s\n", gitTag, gitCommit)

	countdown := countdownData{
		active: false,
	}
	engine.countdown = &countdown
	countdown2 := countdownData{
		active: false,
	}
	engine.countdown2 = &countdown2
	countup := countdownData{
		active: false,
	}
	engine.countup = &countup

	// Setup the OSC listener and feedback
	if !options.DisableOSC {
		engine.oscServer = osc.Server{
			Addr: options.ListenAddr,
		}
		engine.clockServer = MakeServer(&engine.oscServer)
		log.Printf("OSC control: listening on %v", engine.oscServer.Addr)
		go engine.runOSC()

		// process osc commands
		go engine.listen()

		if options.DisableFeedback {
			engine.oscDests = nil
			log.Printf("OSC feedback disabled")
		} else {
			// OSC feedback
			engine.oscDests = initFeedback(options.Connect)
		}
	} else {
		log.Printf("OSC control and feedback disabled.\n")
	}

	// Time zones
	tz, err := time.LoadLocation(options.Timezone)
	if err != nil {
		return nil, err
	}
	engine.timeZone = tz

	// Led flash cycle
	// Setting the interval to 0 disables
	if options.Flash > 0 {
		engine.flasher = time.NewTicker(time.Duration(options.Flash) * time.Millisecond)
		go engine.flash()
	}

	return &engine, nil
}

func (engine *Engine) runOSC() {
	err := engine.oscServer.ListenAndServe()
	if err != nil {
		panic(err)
	}
}

// Listen for OSC messages
func (engine *Engine) listen() {
	oscChan := engine.clockServer.Listen()
	tallyTimer := timer.NewTimer(engine.timeout)

	for {
		select {
		case message := <-oscChan:
			// New OSC message received
			debug.Printf("Got new osc data: %v\n", message)
			switch message.Type {
			case "count":
				msg := message.CountMessage
				engine.TallyRed = uint8(msg.ColorRed)
				engine.TallyGreen = uint8(msg.ColorGreen)
				engine.TallyBlue = uint8(msg.ColorBlue)
				engine.Tally = fmt.Sprintf("%.1s%02d%.1s", msg.Symbol, msg.Count, msg.Unit)
				engine.oscTally = true
				tallyTimer.Reset(engine.timeout)
			case "display":
				msg := message.DisplayMessage
				engine.TallyRed = uint8(msg.ColorRed)
				engine.TallyGreen = uint8(msg.ColorGreen)
				engine.TallyBlue = uint8(msg.ColorBlue)
				engine.Tally = fmt.Sprintf("%-.4s", msg.Text)
				engine.oscTally = true
				tallyTimer.Reset(engine.timeout)
			case "countdownStart":
				msg := message.CountdownMessage
				engine.StartCountdown(time.Duration(msg.Seconds) * time.Second)
			case "countdownStart2":
				msg := message.CountdownMessage
				engine.StartCountdown2(time.Duration(msg.Seconds) * time.Second)
			case "countdownModify":
				msg := message.CountdownMessage
				engine.ModifyCountdown(time.Duration(msg.Seconds) * time.Second)
			case "countdownModify2":
				msg := message.CountdownMessage
				engine.ModifyCountdown2(time.Duration(msg.Seconds) * time.Second)
			case "countdownStop":
				engine.StopCountdown()
			case "countdownStop2":
				engine.StopCountdown2()
			case "pause":
				engine.Pause()
			case "resume":
				engine.Resume()
			case "countup":
				engine.StartCountup()
			case "kill":
				engine.Kill()
			case "normal":
				engine.Normal()
			case "secondsOff":
				engine.displaySeconds = false
			case "secondsOn":
				engine.displaySeconds = true
			case "setTime":
				engine.setTime(message.Data)
			}
			// We have received a osc command, so stop the version display
			engine.initialized = true
		case <-tallyTimer.C:
			// OSC message timeout
			engine.Tally = ""
			engine.oscTally = false
		}
	}
}

// Send the clock state as /clock/state
func (engine *Engine) sendState() error {
	if engine.oscDests == nil {
		// No osc connection
		return nil
	}
	hours := engine.Hours
	minutes := engine.Minutes
	seconds := engine.Seconds

	if seconds == "" {
		hours = ""
		minutes = engine.Hours
		seconds = engine.Minutes
	}
	mode := engine.mode
	pause := int32(0)

	if engine.paused {
		pause = 1
	}

	packet := osc.NewMessage("/clock/state", int32(mode), hours, minutes, seconds, engine.Tally, pause)

	data, err := packet.MarshalBinary()
	if err != nil {
		return err
	}
	engine.oscDests.Write(data)
	return nil
}

func (engine *Engine) flash() {
	for range engine.flasher.C {
		engine.flashLeds = !engine.flashLeds
	}
}

/*
 * Display update requests
 */

// Update Hours, Minutes and Seconds
func (engine *Engine) Update() {
	switch engine.mode {
	case Normal:
		engine.normalUpdate()
	case Countdown:
		engine.countdownUpdate()
	case Countup:
		engine.countupUpdate()
	case Off:
		engine.Hours = ""
		engine.Minutes = ""
		engine.Seconds = ""
		engine.Leds = 0
		engine.Dots = false
	}

	engine.countdown2Update() // Update secondary countdown if needed

	if !engine.displaySeconds {
		// Clear the seconds field as requested
		engine.Seconds = ""
	}

	if err := engine.sendState(); err != nil {
		log.Printf("Error sending osc state: %v", err)
	}
}

func (engine *Engine) normalUpdate() {
	t := time.Now().In(engine.timeZone)
	engine.Dots = true

	// Check that the rpi has valid time
	if t.Year() > 2000 {
		// We have ntp synced time, so display it
		engine.initialized = true
		engine.Hours = t.Format("15")
		engine.Minutes = t.Format("04")
		engine.Seconds = t.Format("05")
		engine.Leds = t.Second()
	} else if !engine.initialized {
		// No valid time, display version number instead of Time of day
		var major, minor, bugfix int
		_, err := fmt.Sscanf(gitTag, "v%d.%d.%d", &major, &minor, &bugfix)
		if err != nil {
			panic(err)
		}
		engine.Tally = "ver"
		engine.TallyRed = 255
		engine.TallyGreen = 255
		engine.TallyBlue = 255
		engine.Hours = fmt.Sprintf("%2d", major)
		engine.Minutes = fmt.Sprintf("%2d", minor)
		engine.Seconds = fmt.Sprintf("%2d", bugfix)
		engine.Dots = false
		engine.Leds = t.Second()
	} else {
		// We have received a osc command but no valid time, black the fields
		engine.Hours = ""
		engine.Minutes = ""
		engine.Seconds = ""
		engine.Dots = false
		engine.Leds = 0
	}
}

func (engine *Engine) countupUpdate() {
	t := time.Now()
	var diff time.Duration
	if engine.paused {
		diff = engine.countup.left
		if engine.flashLeds {
			engine.Dots = true
		} else {
			engine.Dots = false
		}
	} else {
		diff = t.Sub(engine.countup.target)
		engine.Dots = true
	}

	display := time.Time{}.Add(diff)
	engine.Seconds = ""

	if diff > 0 {
		engine.formatCount(diff)
		engine.Leds = display.Minute()
	} else {
		engine.Hours = "00"
		engine.Minutes = "00"
		engine.Leds = 59
	}
}

func (engine *Engine) countdownUpdate() {
	var diff time.Duration
	if engine.paused {
		diff = engine.countdown.left
		if engine.flashLeds {
			engine.Dots = true
		} else {
			engine.Dots = false
		}
	} else {
		t := time.Now()
		diff = engine.countdown.target.Sub(t)
		engine.Dots = true
	}
	engine.Seconds = ""

	// Main countdown
	if diff > 0 {
		engine.formatCount(diff)
		progress := (float64(diff) / float64(engine.countdown.duration))
		if progress >= 1 {
			progress = 1
		} else if progress < 0 {
			progress = 0
		}
		engine.Leds = int(math.Floor(progress * 59))
	} else {
		engine.Leds = 59
		if engine.flashLeds {
			engine.Hours = "00"
			engine.Minutes = "00"
		} else {
			engine.Hours = ""
			engine.Minutes = ""
		}
	}
}

// Secondary countdown, lower priority than Tally messages
func (engine *Engine) countdown2Update() {
	if !engine.initialized {
		// No valid time and no osc messages received
		return
	}

	var diff2 time.Duration
	if engine.paused {
		diff2 = engine.countdown2.left
	} else {
		t := time.Now()
		diff2 = engine.countdown2.target.Sub(t).Truncate(time.Second)
	}
	if !engine.oscTally && !engine.countdown2.active {
		// Clear the countdown display on stop
		engine.Tally = ""
	} else if !engine.oscTally && engine.countdown2.active {
		if diff2 > 0 {
			engine.formatCount2(diff2)
		} else {
			if engine.flashLeds {
				engine.Tally = "↓00"
			} else {
				engine.Tally = ""
			}
		}
	}
}

func (engine *Engine) formatCount(diff time.Duration) {
	hours := int32(diff.Truncate(time.Hour).Hours())
	minutes := int32(diff.Truncate(time.Minute).Minutes()) - (hours * 60)
	secs := int32(diff.Truncate(time.Second).Seconds()) - (((hours * 60) + minutes) * 60)

	if hours > 99 {
		engine.Hours = "++"
		engine.Minutes = "++"
	} else if hours != 0 {
		engine.Hours = fmt.Sprintf("%02d", hours)
		engine.Minutes = fmt.Sprintf("%02d", minutes)
		engine.Seconds = fmt.Sprintf("%02d", secs)
	} else {
		engine.Hours = fmt.Sprintf("%02d", minutes)
		engine.Minutes = fmt.Sprintf("%02d", secs)
	}
}

func (engine *Engine) formatCount2(diff time.Duration) {
	if !engine.oscTally {
		// osc tally messages take priority
		secs := int64(diff.Truncate(time.Second).Seconds())

		for _, unit := range clockUnits {
			if secs/int64(unit.seconds) >= 100 {
				continue
			}
			engine.TallyRed = engine.cd2Red
			engine.TallyGreen = engine.cd2Green
			engine.TallyBlue = engine.cd2Blue
			count := secs / int64(unit.seconds)
			if engine.paused {
				engine.Tally = fmt.Sprintf(" %02d%1s", count, unit.unit)
			} else {
				engine.Tally = fmt.Sprintf("↓%02d%1s", count, unit.unit)
			}
			return
		}
	}
}

/*
 * OSC Message handlers
 */

// StartCountdown starts a primary countdown timer
func (engine *Engine) StartCountdown(timer time.Duration) {
	cd := countdownData{
		target:   time.Now().Add(timer).Truncate(time.Second),
		duration: timer,
		left:     timer,
		active:   true,
	}
	engine.countdown = &cd
	engine.mode = Countdown
}

// StartCountdown2 starts a secondary countdown timer
func (engine *Engine) StartCountdown2(timer time.Duration) {
	cd := countdownData{
		target:   time.Now().Add(timer).Truncate(time.Second),
		duration: timer,
		left:     timer,
		active:   true,
	}
	engine.countdown2 = &cd
}

// StartCountup starts counting time up from this moment
func (engine *Engine) StartCountup() {
	cd := countdownData{
		target: time.Now().Truncate(time.Second),
		active: true,
	}
	engine.countup = &cd
	engine.mode = Countup
}

// Normal returns main display to normal clock
func (engine *Engine) Normal() {
	engine.mode = Normal
	// engine.countdown2.active = false
}

// ModifyCountdown adds or removes time from primary countdown
func (engine *Engine) ModifyCountdown(delta time.Duration) {
	if engine.mode == Countdown {
		cd := countdownData{
			target:   engine.countdown.target.Add(delta),
			duration: engine.countdown.duration + delta,
			left:     engine.countdown.left + delta,
		}
		engine.countdown = &cd
	}
}

// ModifyCountdown2 adds or removes time from primary countdown
func (engine *Engine) ModifyCountdown2(delta time.Duration) {
	if engine.countdown2.active {
		cd := countdownData{
			target:   engine.countdown2.target.Add(delta),
			duration: engine.countdown2.duration + delta,
			left:     engine.countdown2.left + delta,
			active:   engine.countdown2.active,
		}
		engine.countdown2 = &cd
	}
}

// StopCountdown stops the primary countdown
func (engine *Engine) StopCountdown() {
	if engine.mode == Countdown {
		cd := countdownData{
			target:   time.Now(),
			duration: time.Millisecond,
			left:     time.Millisecond,
			active:   false,
		}
		engine.countdown = &cd
		engine.mode = Off
	}
}

// StopCountdown2 stops the secondary countdown
func (engine *Engine) StopCountdown2() {
	cd := countdownData{
		target:   time.Now(),
		duration: time.Millisecond,
		left:     time.Millisecond,
		active:   false,
	}
	engine.countdown2 = &cd
}

// Kill blanks the clock display
func (engine *Engine) Kill() {
	engine.mode = Off
	engine.countdown2.active = false
}

// Pause pauses both countdowns
func (engine *Engine) Pause() {
	if !engine.paused {
		t := time.Now()
		engine.countdown.left = engine.countdown.target.Sub(t).Truncate(time.Second)
		engine.countdown2.left = engine.countdown2.target.Sub(t).Truncate(time.Second)
		engine.countup.left = t.Sub(engine.countup.target).Truncate(time.Second)
		engine.paused = true
	}
}

// Resume resumes both countdowns if they have been paused
func (engine *Engine) Resume() {
	if engine.paused {
		t := time.Now()
		engine.countdown.target = t.Add(engine.countdown.left).Truncate(time.Second)
		engine.countdown2.target = t.Add(engine.countdown2.left).Truncate(time.Second)
		engine.countup.target = t.Add(-engine.countup.left).Truncate(time.Second)
		engine.paused = false
	}
}

// DisplaySeconds returns true if the clock should display seconds
func (engine *Engine) DisplaySeconds() bool {
	return engine.displaySeconds
}

// TimeOfDay returns true if the clock is showing time of day, false otherwise
func (engine *Engine) TimeOfDay() bool {
	return engine.mode == Normal
}

func (engine *Engine) setTime(time string) {
	debug.Printf("Set time: %#v", time)
	_, lookErr := exec.LookPath("date")
	if lookErr != nil {
		debug.Printf("Date binary not found, cannot set system date: %s\n", lookErr.Error())
		return
	}
	// Validate the received time
	match, _ := regexp.MatchString("^(2[0-3]|[01][0-9]):([0-5][0-9]):([0-5][0-9])$", time)
	if match {
		// Set the system time
		dateString := fmt.Sprintf("2019-01-01 %s", time)
		tzString := fmt.Sprintf("TZ=%s", engine.timeZone.String())
		debug.Printf("Setting system date to: %s\n", dateString)
		args := []string{"--set", dateString}
		cmd := exec.Command("date", args...)
		cmd.Env = os.Environ()
		cmd.Env = append(cmd.Env, tzString)
		cmd.Run()
	} else {
		debug.Printf("Invalid time provided: %v\n", time)
	}
}
